### PR TITLE
refactor: type legacy credential shape scans (#417)

### LINKED ISSUE #417: Typed-boundary discipline for legacy plaintext credential shape detection
## Context

The plaintext-credential startup guard added in #413 (`refactor/remove-legacy-compat-code`) detects WhatsApp/pairing/Baileys-shaped JSON files via raw string-key matching against hand-maintained constants:

- `WHATSAPP_LEGACY_CREDENTIAL_JSON_KEYS` — `src/credentials/mod.rs:56-81` (23 keys)
- `PAIRING_LEGACY_CREDENTIAL_JSON_KEYS` — `src/credentials/mod.rs:82-101` (18 keys)

The match logic is `object.keys().any(|k| LIST.contains(&k.as_str()))` after `serde_json::from_str::<Value>`.

This violates the project's typed-boundary discipline rule (`.claude/rules/rust-patterns.md`):

> Prefer explicit structs/enums/DTOs over `serde_json::Value`, free-form maps, or string parsing at important boundaries.
>
> Treat the following as important boundaries by default: ... security-sensitive projections/redaction layers ... If review churn repeats on the same serialization/projection seam, treat that as evidence the boundary is under-typed and widen to the owning typed abstraction instead of adding another sanitizer, allowlist, or string match.

The credential-shape detection is a security-relevant seam (its purpose is to reject plaintext credential files at startup), and the current implementation is exactly the \"add another allowlist\" pattern the rule warns against.

## Why it matters

1. **Genericness of allowlist keys.** `PAIRING_LEGACY_CREDENTIAL_JSON_KEYS` includes `\"key\"`, `\"session\"`, `\"store\"`, `\"identity\"`, `\"phone\"`, `\"contacts\"`, `\"token\"`, `\"secret\"` — generic keys that incidental JSON files may contain. Any `*-pairing.json` whose root has one of these keys is treated as a credential and rejected. The carve-out for incidental files is currently covered by only one test (`{\"message\": \"debug note\"}`).

2. **Drift risk.** When upstream Baileys or the pairing schema renames a field, the heuristic silently rots — a real credential file may stop being detected. Typed deserialization fails loudly on schema drift.

3. **Audit trail.** Hand-maintained string lists at security boundaries make it harder to reason about what the guard rejects vs. accepts. A typed `LegacyCredentialEnvelope` with canonical fields makes the security contract explicit.

## Proposal

Replace the string-allowlist heuristic with typed deserialization:

```rust
#[derive(Deserialize)]
struct WhatsappLegacyCredentialShape {
    // Canonical Baileys credential fields. Optional so partial files
    // still match; deny_unknown_fields = false so unrelated keys are
    // ignored rather than blocking a real credential.
    noise_key:           Option<serde_json::Value>,
    signed_identity_key: Option<serde_json::Value>,
    registration_id:     Option<u64>,
    // ... small canonical set
}

fn whatsapp_json_has_credential_shape(json: &str) -> bool {
    serde_json::from_str::<WhatsappLegacyCredentialShape>(json)
        .map(|s| s.noise_key.is_some() || s.signed_identity_key.is_some() || s.registration_id.is_some())
        .unwrap_or(false)
}
```

Pairing files have a richer shape (string / array / object branches at the top level — see `pairing_json_has_credential_shape` at `src/credentials/mod.rs:309-318`); a typed `LegacyPairingEnvelope` enum can encode that.

## Acceptance criteria

- The two `*_LEGACY_CREDENTIAL_JSON_KEYS` constants in `src/credentials/mod.rs` are deleted.
- Both `whatsapp_json_has_credential_shape` and `pairing_json_has_credential_shape` operate on typed structs/enums.
- Existing tests in `credentials::tests` still pass with no semantic change to the guard's accept/reject behavior on real Baileys/pairing payloads.
- The typed shapes use a small canonical key set (2–4 fields each) — no transcription of the full 23+18 string list into struct fields.

## Out of scope

- Changing the fail-closed posture (`true` returned on read/parse failure stays as-is).
- `is_legacy_whatsapp_plaintext_file` filename-based gate (separate concern).

### PR DESCRIPTION
**Pull request by puremachinery**
## Summary

- replace legacy WhatsApp and pairing credential JSON key allowlists with typed serde shapes
- keep fail-closed startup scanning behavior while reducing generic-key false positives for pairing files
- add coverage for canonical WhatsApp fields and generic pairing-key incidental JSON

Fixes #417.

## Validation

- `scripts/cargo-serial nextest run -p carapace --filter-expr 'test(credential) | test(plaintext_credential_guard) | test(pairing_credential_shape) | test(whatsapp_credential_shape)'`
- `scripts/cargo-serial clippy --all-targets -- -D warnings`

### CODE REVIEW

File: src/credentials/mod.rs [hunk 1]
```diff
@@ -150,6 +104,194 @@ enum CredentialShapeScan {
...
+    }
+}
+
+fn deserialize_optional_shape_field<'de, D, T>(deserializer: D) -> Result<Option<T>, D::Error>
+where
+    D: Deserializer<'de>,
+    T: serde::de::DeserializeOwned,
+{
+    let value = Value::deserialize(deserializer)?;
+    Ok(serde_json::from_value(value).ok())
```
[the remark this answers is not in the document]
Comment (puremachinery) [reply src/credentials/mod.rs]: Outdated/addressed: the Value round-trip helper was removed. The current shape scan deserializes typed fields directly (for example Option<PresentJsonField>) instead of cloning through serde_json::Value.

File: src/credentials/mod.rs [hunk 2]
```diff
@@ -150,6 +104,194 @@ enum CredentialShapeScan {
...
+
+struct JsonArray;
+
+impl<'de> Deserialize<'de> for JsonArray {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let _ = Vec::<serde::de::IgnoredAny>::deserialize(deserializer)?;
+        Ok(Self)
```
[the remark this answers is not in the document]
Comment (puremachinery) [reply src/credentials/mod.rs]: Addressed: JsonArray and NonEmptyJsonArray now use custom serde visitors that consume sequences via IgnoredAny without allocating Vecs.

File: src/credentials/mod.rs [hunk 3]
```diff
@@ -150,6 +104,194 @@ enum CredentialShapeScan {
...
+            || self.pairing.as_ref().is_some_and(pairing_value_is_present)
+            || self
+                .credential
+                .as_ref()
+                .is_some_and(pairing_value_is_present)
+            || self
+                .credentials
+                .as_ref()
+                .is_some_and(pairing_value_is_present)
+    }
```
[the remark this answers is not in the document]
Comment (puremachinery) [reply src/credentials/mod.rs]: Outdated: envelope keys now use PresentJsonField presence semantics, so values like {"pairing": {}} are still classified as credential-shaped. The canonical-fields test includes that case.

File: src/credentials/mod.rs [hunk 4]
```diff
@@ -476,16 +618,9 @@ fn whatsapp_plaintext_file_credential_scan(path: &Path) -> CredentialShapeScan {
...
-        return false;
-    };
-
-    object.keys().any(|key| {
-        WHATSAPP_LEGACY_CREDENTIAL_JSON_KEYS.contains(&key.as_str())
-            || (key == "type"
-                && object.get("type").and_then(Value::as_str) == Some("Buffer")
-                && object.get("data").is_some_and(Value::is_array))
-    })
+    LegacyWhatsappBufferShape::deserialize(value).is_ok_and(|shape| shape.has_credential_shape())
```
[the remark this answers is not in the document]
Comment (puremachinery) [reply src/credentials/mod.rs]: Outdated: the WhatsApp mapping now covers all 24 former keys across the typed sub-shapes, and the per-field test asserts each mapped field plus snake_case aliases where applicable.

File: src/credentials/mod.rs [hunk 5]
```diff
@@ -150,6 +104,194 @@ enum CredentialShapeScan {
...
+                .as_ref()
+                .is_some_and(pairing_value_is_present)
+            || self
+                .credentials
+                .as_ref()
+                .is_some_and(pairing_value_is_present)
+    }
+}
+
+fn pairing_value_is_present(value: &PairingCredentialValue) -> bool {
```
[the remark this answers is not in the document]
Comment (puremachinery) [reply src/credentials/mod.rs]: Addressed: pairing_value_is_present was removed. Pairing object shape detection now uses direct Option::is_some checks on fields whose deserialization already establishes presence.

File: src/credentials/mod.rs [hunk 6]
```diff
@@ -150,6 +104,194 @@ enum CredentialShapeScan {
...
+    kind: LegacyWhatsappBufferKind,
+    data: JsonArray,
+}
+
+impl LegacyWhatsappBufferShape {
+    fn has_credential_shape(&self) -> bool {
+        let _ = &self.data;
+        matches!(self.kind, LegacyWhatsappBufferKind::Buffer)
+    }
+}
```
[the remark this answers is not in the document]
Comment (puremachinery) [reply src/credentials/mod.rs]: Addressed: LegacyWhatsappBufferShape no longer has a redundant has_credential_shape method or data no-op. The caller treats successful typed buffer deserialization as the shape check.

File: src/credentials/mod.rs [hunk 7]
```diff
@@ -150,6 +104,194 @@ enum CredentialShapeScan {
...
+        PairingCredentialValue::Array(value) => {
+            let _ = value;
+            true
+        }
+        PairingCredentialValue::Object(value) => {
+            let _ = value;
+            true
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (puremachinery) [reply src/credentials/mod.rs]: Addressed: pairing_value_is_present and the is_some_and indirection were removed; has_credential_shape now directly checks whether the typed fields are present.

File: src/credentials/mod.rs [hunk 8]
```diff
@@ -476,16 +618,9 @@ fn whatsapp_plaintext_file_credential_scan(path: &Path) -> CredentialShapeScan {
...
-
-    object.keys().any(|key| {
-        WHATSAPP_LEGACY_CREDENTIAL_JSON_KEYS.contains(&key.as_str())
-            || (key == "type"
-                && object.get("type").and_then(Value::as_str) == Some("Buffer")
-                && object.get("data").is_some_and(Value::is_array))
-    })
+    LegacyWhatsappBufferShape::deserialize(value).is_ok_and(|shape| shape.has_credential_shape())
+        || LegacyWhatsappCredentialShape::deserialize(value)
+            .is_ok_and(|shape| shape.has_credential_shape())
```
[the remark this answers is not in the document]
Comment (puremachinery) [reply src/credentials/mod.rs]: Addressed: whatsapp_json_has_credential_shape now uses successful LegacyWhatsappBufferShape deserialization directly for Buffer-shaped values, and the credential object path uses is_ok_and with has_credential_shape.

File: src/credentials/mod.rs [hunk 9]
```diff
@@ -495,13 +630,17 @@ fn pairing_plaintext_file_credential_scan(path: &Path) -> CredentialShapeScan {
...
+            let _ = value;
+            true
+        }
+        Ok(LegacyPairingShape::Array(value)) => {
+            let _ = value;
+            true
+        }
+        Ok(LegacyPairingShape::Object(shape)) => shape.has_credential_shape(),
+        Err(_) => false,
     }
```
[the remark this answers is not in the document]
Comment (puremachinery) [reply src/credentials/mod.rs:779]: Addressed in 4c784db by simplifying the match to Ok(LegacyPairingShape::String(_) | LegacyPairingShape::Array(_)) => true.

File: src/credentials/mod.rs [hunk 10]
```diff
@@ -150,6 +104,194 @@ enum CredentialShapeScan {
...
+    credentials: Option<PairingCredentialValue>,
+}
+
+impl LegacyPairingObjectShape {
+    fn has_credential_shape(&self) -> bool {
+        self.pairing_code.is_some()
+            || self.pairing.as_ref().is_some_and(pairing_value_is_present)
+            || self
+                .credential
+                .as_ref()
```
[the remark this answers is not in the document]
Comment (puremachinery) [reply src/credentials/mod.rs]: Outdated: token-style pairing payloads are preserved as credential signals. LegacyPairingCredentialRecordShape includes token, and the integration test writes a token-keyed *-pairing.json that is rejected end-to-end.

File: src/credentials/mod.rs [hunk 11]
```diff
@@ -150,6 +104,148 @@ enum CredentialShapeScan {
...
+    D: Deserializer<'de>,
+    T: serde::de::DeserializeOwned,
+{
+    let value = Value::deserialize(deserializer)?;
+    Ok(serde_json::from_value(value).ok())
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct LegacyWhatsappCredentialShape {
```
[the remark this answers is not in the document]
Comment (puremachinery) [reply src/credentials/mod.rs]: Outdated: the current WhatsApp typed shapes preserve all 24 former keys across the sub-shapes. The exhaustive per-field test covers the camelCase keys and snake_case aliases.

File: src/credentials/mod.rs [hunk 12]
```diff
@@ -150,6 +104,148 @@ enum CredentialShapeScan {
...
+            Value::Object(object) if !object.is_empty() => Ok(Self),
+            _ => Err(<D::Error as serde::de::Error>::custom(
+                "expected non-empty pairing credential value",
+            )),
+        }
+    }
+}
+
+#[derive(Deserialize)]
+struct LegacyPairingObjectShape {
```
[the remark this answers is not in the document]
Comment (puremachinery) [reply src/credentials/mod.rs]: Addressed/clarified: token, secret, and jid are preserved credential-record anchors. session and other generic operational keys are intentionally ignored to reduce false positives; 4c784db adds an integration test for files containing only generic keys.

File: src/credentials/mod.rs [hunk 13]
```diff
@@ -150,6 +104,148 @@ enum CredentialShapeScan {
...
+#[derive(Deserialize)]
+struct LegacyWhatsappBufferShape {
+    #[serde(rename = "type")]
+    kind: LegacyWhatsappBufferKind,
+    data: JsonArray,
+}
+
+impl LegacyWhatsappBufferShape {
+    fn has_credential_shape(&self) -> bool {
+        let _ = &self.data;
```
[the remark this answers is not in the document]
Comment (puremachinery) [reply src/credentials/mod.rs]: Addressed: the dead no-op was removed. The buffer shape now relies on typed deserialization with _kind/_data fields and the caller checks deserialize(value).is_ok().

File: src/credentials/mod.rs [hunk 14]
```diff
@@ -150,6 +104,318 @@ enum CredentialShapeScan {
...
+        self.pairing_code.is_some()
+            || self.pairing.is_some()
+            || self.credential.is_some()
+            || self.credentials.is_some()
+    }
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct LegacyPairingCredentialRecordShape {
```
[the remark this answers is not in the document]
Comment (puremachinery) [reply src/credentials/mod.rs:365]: Addressed: deviceId is preserved in LegacyPairingCredentialRecordShape with both camelCase and device_id alias support, and the canonical-fields test covers both forms.

File: src/credentials/mod.rs [hunk 15]
```diff
@@ -1974,6 +2250,42 @@ mod tests {
             .expect("incidental pairing files should not block startup");
     }
 
+    #[test]
+    fn test_pairing_credential_shape_ignores_generic_keys_without_envelope() {
```
[the remark this answers is not in the document]
Comment (puremachinery) [reply src/credentials/mod.rs]: Resolved as positive coverage rather than negative coverage: deviceId is not treated as a generic ignored key. The canonical-fields test asserts both deviceId and device_id are detected.

File: src/credentials/mod.rs [hunk 16]
```diff
@@ -2084,6 +2396,38 @@ mod tests {
...
+            "signed_identity_key": {"public": {"type": "Buffer", "data": [2]}},
+            "registration_id": 42
+        });
+
+        assert!(whatsapp_json_has_credential_shape(&camel_case));
+        assert!(whatsapp_json_has_credential_shape(&snake_case));
+    }
+
+    #[test]
+    fn test_whatsapp_credential_shape_detects_session_and_key_material_fields() {
```
[the remark this answers is not in the document]
Comment (puremachinery) [reply src/credentials/mod.rs]: Addressed: test_whatsapp_credential_shape_detects_each_mapped_field now covers every mapped WhatsApp field individually, including the snake_case aliases that exist for the renamed fields.

File: src/credentials/mod.rs [hunk 17]
```diff
@@ -1936,7 +2215,7 @@ mod tests {
         std::fs::create_dir_all(&credentials_dir).unwrap();
         let pairing_path = credentials_dir.join("telegram-pairing.json");
         let allow_from_path = credentials_dir.join("telegram-allowFrom.json");
-        std::fs::write(&pairing_path, r#"{"token":"secret"}"#).unwrap();
+        std::fs::write(&pairing_path, r#"{"pairingCode":"123-456"}"#).unwrap();
```
[the remark this answers is not in the document]
Comment (puremachinery) [reply src/credentials/mod.rs:2230]: Addressed: the integration test now includes both a pairingCode fixture and a token-keyed *-pairing.json fixture, so token detection is covered end-to-end as well as at unit level.

File: src/credentials/mod.rs [hunk 18]
```diff
@@ -1974,6 +2262,43 @@ mod tests {
...
+            serde_json::json!({"pairingCode": "123-456"}),
+            serde_json::json!({"pairing_code": "123-456"}),
+            serde_json::json!({"pairing": {"code": "123-456"}}),
+            serde_json::json!({"pairing": {}}),
+            serde_json::json!({"credential": "pairing-secret"}),
+            serde_json::json!({"credentials": ["pairing-secret"]}),
+            serde_json::json!({"token": "pairing-secret"}),
+            serde_json::json!({"secret": "pairing-secret"}),
+            serde_json::json!({"jid": "debug@example.invalid"}),
+            serde_json::json!({"clientId": "debug-client"}),
```
[the remark this answers is not in the document]
Comment (puremachinery) [reply src/credentials/mod.rs:2322]: Addressed: test_pairing_credential_shape_detects_canonical_object_fields now asserts both client_id and device_id aliases in addition to the camelCase forms.

File: src/credentials/mod.rs [hunk 19]
```diff
@@ -2084,6 +2409,56 @@ mod tests {
...
+            "signed_identity_key": {"public": {"type": "Buffer", "data": [2]}},
+            "registration_id": 42
+        });
+
+        assert!(whatsapp_json_has_credential_shape(&camel_case));
+        assert!(whatsapp_json_has_credential_shape(&snake_case));
+    }
+
+    #[test]
+    fn test_whatsapp_credential_shape_detects_each_mapped_field() {
```
[the remark this answers is not in the document]
Comment (puremachinery) [reply src/credentials/mod.rs:2462]: Addressed: the WhatsApp per-field test now includes the snake_case aliases for the aliased fields, in addition to all camelCase canonical keys.

File: src/credentials/mod.rs [hunk 20]
```diff
@@ -495,13 +767,14 @@ fn pairing_plaintext_file_credential_scan(path: &Path) -> CredentialShapeScan {
...
-    match value {
-        Value::String(value) => !value.trim().is_empty(),
-        Value::Array(items) => !items.is_empty(),
-        Value::Object(object) => object
-            .keys()
-            .any(|key| PAIRING_LEGACY_CREDENTIAL_JSON_KEYS.contains(&key.as_str())),
-        _ => false,
+    match LegacyPairingShape::deserialize(value) {
+        Ok(LegacyPairingShape::String(_)) | Ok(LegacyPairingShape::Array(_)) => true,
+        Err(_) => {
```
[the remark this answers is not in the document]
Comment (puremachinery) [reply src/credentials/mod.rs:773]: Addressed by making the tradeoff explicit in code and tests: pairing object detection keeps envelope keys plus token/secret/jid/clientId/deviceId anchors, while generic fields like session/store/identity remain ignored. 4c784db adds an integration test for a *-pairing.json containing only generic keys.

File: src/credentials/mod.rs [hunk 21]
```diff
@@ -2084,6 +2415,76 @@ mod tests {
...
+            "signed_identity_key": {"public": {"type": "Buffer", "data": [2]}},
+            "registration_id": 42
+        });
+
+        assert!(whatsapp_json_has_credential_shape(&camel_case));
+        assert!(whatsapp_json_has_credential_shape(&snake_case));
+    }
+
+    #[test]
+    fn test_whatsapp_credential_shape_detects_each_mapped_field() {
```
[the remark this answers is not in the document]
Comment (puremachinery) [reply src/credentials/mod.rs:2462]: Addressed in 4c784db with test_whatsapp_credential_shape_ignores_objects_without_mapped_fields, covering empty and incidental JSON objects that should not be classified as WhatsApp credentials.

**Comment by puremachinery** [comment]
Addressed the PR 420 aggregate and inline review findings in 4c784db. Changes include simplifying the pairing shape match, adding an integration test that documents the intentionally ignored generic pairing-key boundary, and adding WhatsApp false-negative coverage for incidental JSON objects. Existing branch state also preserves deviceId/client_id/device_id, token-keyed pairing integration coverage, all WhatsApp mapped fields, and snake_case alias coverage. Focused credential-shape tests and clippy pass locally, and all open review threads have been replied to and resolved.

File: src/credentials/mod.rs [hunk 22]
```diff
@@ -150,6 +104,327 @@ enum CredentialShapeScan {
...
+            || self.sync.has_credential_shape()
+    }
+}
+
+#[derive(Deserialize)]
+struct LegacyWhatsappBufferShape {
+    #[serde(rename = "type")]
+    _kind: LegacyWhatsappBufferKind,
+    #[serde(rename = "data")]
+    _data: JsonArray,
```
[the remark this answers is not in the document]
Comment (puremachinery) [reply src/credentials/mod.rs]: Fixed in feb27a6. `LegacyWhatsappBufferShape` now requires `data` to deserialize as `NonEmptyJsonArray`, and `test_whatsapp_credential_shape_ignores_objects_without_mapped_fields` covers the empty-Buffer false-positive case. The old generic `JsonArray` visitor became dead after that change and was removed.

**Comment by puremachinery** [comment]
Addressed the latest PR 420 review round in feb27a6.

- `LegacyWhatsappBufferShape` now requires non-empty `data`, with regression coverage for `{type: Buffer, data: []}` being ignored.
- The obsolete generic `JsonArray` visitor was removed after the Buffer shape moved to `NonEmptyJsonArray`.
- `LegacyPairingCredentialRecordShape` now explicitly lists the broad previously matched keys that are intentionally excluded: allowFrom, allowedFrom, allowlist, contacts, identity, key, phone, senders, session, and store.
- I did not expand the advisory flatten-routing test in this pass; the current exhaustive field-detection tests and disjoint-field comment cover the present behavior, and the new review round had concrete in-scope changes on precision and auditability.

Validation: focused credential-shape nextest filter passed, 7/7; `scripts/cargo-serial clippy --all-targets -- -D warnings` passed.

File: src/credentials/mod.rs [hunk 23]
```diff
@@ -150,6 +104,299 @@ enum CredentialShapeScan {
...
+        self.pairing_code.is_some()
+            || self.pairing.is_some()
+            || self.credential.is_some()
+            || self.credentials.is_some()
+    }
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct LegacyPairingCredentialRecordShape {
```
[the remark this answers is not in the document]
Comment (puremachinery) [reply src/credentials/mod.rs:365]: Fixed in 3448b4d. The typed pairing shape now preserves the old fail-closed scan anchors for allowFrom, allowedFrom, allowlist, contacts, identity, key, phone, senders, session, and store. I also inverted the prior generic-key tests so a *-pairing.json file with only those broad legacy keys blocks startup instead of silently passing.

File: src/credentials/mod.rs [hunk 24]
```diff
@@ -476,16 +723,13 @@ fn whatsapp_plaintext_file_credential_scan(path: &Path) -> CredentialShapeScan {
...
-        return false;
-    };
-
-    object.keys().any(|key| {
-        WHATSAPP_LEGACY_CREDENTIAL_JSON_KEYS.contains(&key.as_str())
-            || (key == "type"
-                && object.get("type").and_then(Value::as_str) == Some("Buffer")
-                && object.get("data").is_some_and(Value::is_array))
-    })
+    if LegacyWhatsappBufferShape::deserialize(value).is_ok() {
```
[the remark this answers is not in the document]
Comment (puremachinery) [reply src/credentials/mod.rs:755]: Fixed in 3448b4d. `test_whatsapp_credential_shape_detects_canonical_fields` now has a direct positive assertion for a top-level non-empty `{type: "Buffer", data: [1]}` value, complementing the empty-data negative assertion from the prior round.

**Comment by puremachinery** [comment]
Addressed the latest PR 420 review round in 3448b4d.

- The pairing scanner now preserves the old fail-closed anchors as typed fields: allowFrom, allowedFrom, allowlist, contacts, identity, key, phone, senders, session, and store.
- The previous generic-key tests are now fail-closed coverage for broad legacy pairing keys, including startup rejection for a *-pairing.json file using only those keys.
- Added a direct positive assertion for the top-level non-empty WhatsApp Buffer path.
- I did not expand the advisory flatten-disjointness test here. The present field coverage still exercises every mapped field, and the concrete regression risk in this round was the pairing scan narrowing, which is now removed.

Validation: focused credential-shape nextest filter passed, 7/7; `scripts/cargo-serial clippy --all-targets -- -D warnings` passed.
